crates/sshx/src/controller.rs

```diff
@@ -46,25 +47,51 @@ pub struct Controller {
 
 impl Controller {
     /// Construct a new controller, connecting to the remote server.
-    pub async fn new(origin: &str, name: &str, runner: Runner) -> Result<Self> {
+    pub async fn new(
+        origin: &str,
+        name: &str,
+        runner: Runner,
+        enable_readers: bool,
+    ) -> Result<Self> {
         debug!(%origin, "connecting to server");
         let encryption_key = rand_alphanumeric(14); // 83.3 bits of entropy
+        let write_password = rand_alphanumeric(14); // 83.3 bits of entropy
 
         let kdf_task = {
             let encryption_key = encryption_key.clone();
             task::spawn_blocking(move || Encrypt::new(&encryption_key))
         };
+
+        let kdf_write_password_task = {
```

Comment:
small change: Can we move the `if enable_readers` branch up one level so we don't need to do argon2 construction if it's not used? It's moved into a separate thread since it might take some time. (couple hundred millis)